// @devseed-ui/modal/src/styled.js
import styled, { createGlobalStyle, css } from 'styled-components';

import { themeVal, glsp } from '@devseed-ui/theme-provider';
import { Toolbar } from '@devseed-ui/toolbar';

const sizeMapping = {
  small: '32rem',
  medium: '48rem',
  large: '64rem',
  xlarge: '80rem',
  fullscreen: '100%'
};

const renderModalSize = (props = {}) => {
  const { size = 'medium' } = props;
  return css`
    max-width: ${sizeMapping[size]};
  `;
};

export const ModalContents = styled.div`
  display: grid;
  grid-template-columns: 100%;
  grid-template-rows: ${({ size }) =>
    size === 'fullscreen' ? `3rem auto` : `auto`};
  background: ${themeVal('color.surface')};
  border-radius: ${themeVal('shape.rounded')};
  width: 100%;
  margin: ${({ size }) => (size === 'fullscreen' ? 0 : `auto`)};
  box-shadow: ${themeVal('boxShadow.elevationD')};

  /* Size attribute */
  ${renderModalSize}

  > *:last-child {
    margin-bottom: 0;
  }
`;

export const ModalWrapper = styled.section`
  display: flex;
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 9990;
  overflow-y: auto;
  opacity: 1;
  visibility: visible;
  background: ${themeVal('color.base-400a')};
  transform: translate3d(0, 0, 0);
  padding: ${({ size }) => (size === 'fullscreen' ? 0 : glsp(2))};

  &.modal-appear,
  &.modal-enter {
    opacity: 0;
    visibility: hidden;
  }

  &.modal-enter-appear,
  &.modal-enter-active {
    transition: opacity 0.32s ease 0.1s, visibility 0.32s linear 0.1s;
    opacity: 1;
    visibility: visible;
  }

  &.modal-exit {
    opacity: 1;
    visibility: visible;
  }

  &.modal-exit-active {
    transition: opacity 0.32s ease 0.1s, visibility 0.32s linear 0.1s;
    opacity: 0;
    visibility: hidden;
  }
`;

export const BodyUnscrollable = createGlobalStyle`
  ${({ revealed }) =>
    revealed &&
    `
    body {
      overflow-y: hidden;
    }
  `}
`;

export const ModalHeader = styled.header`
  display: flex;
  gap: ${glsp()};
  flex-flow: row nowrap;
  align-items: center;
  background: ${themeVal('color.surface')};
  padding: ${glsp(2)};
  border-radius: ${themeVal('shape.rounded')} ${themeVal('shape.rounded')} 0 0;
`;

export const ModalHeadline = styled.div`
  display: flex;
  flex-flow: column nowrap;
`;

export const ModalToolbar = styled(Toolbar)`
  margin-left: auto;
  padding: ${glsp(1 / 8, 0, 1 / 8, 0)};
`;

export const ModalBody = styled.div`
  padding: 0 ${glsp(2)};

  &:first-child {
    padding-top: ${glsp(2)};
  }

  &:last-child {
    padding-bottom: ${glsp(2)};
  }
`;

export const ModalFooter = styled.footer`
  background: ${themeVal('color.surface')};
  padding: ${glsp(2)};
  border-radius: 0 0 ${themeVal('shape.rounded')} ${themeVal('shape.rounded')};
`;
