// @devseed-ui/animation/src/story-animate.js
import React, { useState } from 'react';
import styled, { css } from 'styled-components';
import { Button } from '@devseed-ui/button';
import { Heading } from '@devseed-ui/typography';
import { themeVal, glsp } from '@devseed-ui/theme-provider';

import { chessBg } from '../../../.storybook/story-utils';

const Square = styled.div`
  width: ${glsp(3)};
  height: ${glsp(3)};
  background-color: ${themeVal('color.primary')};

  ${(props) =>
    props.isAnimating &&
    css`
      animation: ${props.animation} ${props.duration} ${props.easing};
    `}
`;

const AnimationWrapper = styled.div`
  padding: ${glsp()};
  display: grid;
  grid-gap: ${glsp(0.5)};

  ${Button} {
    justify-self: start;
  }

  > div {
    border: 1px solid ${themeVal('color.base-400')};
    padding: ${glsp(2)};
    ${chessBg}
  }
`;

export const Animate = (props) => {
  /* eslint-disable-next-line react/prop-types */
  const { title, animation, duration, easing } = props;

  const [animating, setAnimating] = useState(false);

  const play = () => {
    setAnimating(true);
    setTimeout(() => {
      setAnimating(false);
    }, 1000);
  };

  return (
    <AnimationWrapper>
      <Heading as='h2'>{title}</Heading>
      <Button
        variation='primary-fill'
        size='small'
        disabled={animating}
        onClick={play}
      >
        Play
      </Button>
      <div>
        <Square
          isAnimating={animating}
          animation={animation}
          duration={duration}
          easing={easing}
        />
      </div>
    </AnimationWrapper>
  );
};
