// @devseed-ui/accordion/src/styled.js
import styled from 'styled-components';

import { themeVal, glsp } from '@devseed-ui/theme-provider';
import { CollecticonChevronDownSmall } from '@devseed-ui/collecticons';

export const AccordionFoldSelf = styled.section`
  display: grid;
  background: ${themeVal('color.surface')};
  border-radius: ${themeVal('shape.rounded')};
`;

export const AccordionFoldHeader = styled.header`
  box-shadow: 0 1px 0 0 ${themeVal('color.base-100a')};
`;

export const AccordionFoldTrigger = styled.button`
  display: flex;
  align-items: center;
  padding: ${glsp()};
  text-decoration: none;
  border: 0;
  background: inherit;
  color: inherit;
  width: 100%;
  cursor: pointer;

  ${CollecticonChevronDownSmall} {
    transition: transform 240ms ease-in-out;
    transform: ${({ isExpanded }) =>
      isExpanded ? 'rotate(180deg)' : 'rotate(0deg)'};
    margin-left: auto;
  }
`;

export const AccordionFoldBody = styled.div`
  overflow: hidden;
  transition: ${({ isExpanded }) =>
    isExpanded
      ? 'max-height 320ms ease-in-out, opacity 160ms ease-in-out'
      : 'max-height 320ms ease-in-out, opacity 160ms ease-in-out 160ms'};
  opacity: ${({ isExpanded }) => (isExpanded ? 1 : 0)};
  max-height: ${({ isExpanded }) => (isExpanded ? '400vh' : 0)};
`;

export const AccordionFoldBodyInner = styled.div`
  box-shadow: 0 1px 0 0 ${themeVal('color.base-100a')};
  background: ${themeVal('color.base-50a')};
  padding: ${glsp()};
`;
